mempool/src/pool.rs

```diff
@@ -0,0 +1,330 @@
+        }
+
+        if tx.get_outputs().is_empty() {
+            return Err(TxValidationError::NoOutputs);
+        }
+
+        if tx.is_coinbase() {
+            return Err(TxValidationError::LooseCoinbase);
+        }
+
+        // TODO consier a MAX_MONEY check reminiscent of bitcoin's
+        // TODO consider rejecting non-standard transactions (for some definition of standard)
+        if has_duplicate_entry(tx.get_inputs().iter().map(|input| input.get_outpoint())) {
+            return Err(TxValidationError::DuplicateInputs);
+        }
+
+        if tx.encoded_size() > MAX_BLOCK_SIZE_BYTES {
+            return Err(TxValidationError::ExceedsMaxBlockSize);
+        }
+
+        if self.contains_transaction(&tx.get_id()) {
+            return Err(TxValidationError::TransactionAlreadyInMempool);
+        }
+
+        tx.get_inputs()
+            .iter()
+            .filter_map(|input| self.store.find_conflicting_tx(input.get_outpoint()))
+            .all(|tx| tx.is_replaceable())
+            .then(|| ())
+            .ok_or(TxValidationError::ConflictWithIrreplaceableTransaction)?;
+
+        self.verify_inputs_available(tx)?;
+
+        Ok(())
+    }
+}
+
+impl<C: ChainState + Debug> Mempool<C> for MempoolImpl<C> {
+    fn create(chain_state: C) -> Self {
+        Self {
+            store: MempoolStore::new(),
+            chain_state,
+        }
+    }
+
+    fn new_tip_set(&mut self) -> Result<(), MempoolError> {
+        unimplemented!()
+    }
+    //
+
+    fn add_transaction(&mut self, tx: Transaction) -> Result<(), MempoolError> {
+        // TODO (1). First, we need to decide on criteria for the Mempool to be considered full. Maybe number
+        // of transactions is not a good enough indicator. Consider checking mempool size as well
+        // TODO (2) What to do when the mempool is full. Instead of rejecting Do incoming transaction we probably want to evict a low-score transaction
+        if self.store.txs_by_fee.len() >= MEMPOOL_MAX_TXS {
+            return Err(MempoolError::MempoolFull);
+        }
+        self.validate_transaction(&tx)?;
+        self.store.add_tx(tx.clone())?;
+        self.chain_state.apply_tx(&tx);
+        Ok(())
+    }
+
+    fn get_all(&self) -> Vec<&Transaction> {
+        self.store.txs_by_fee.iter().map(|mempool_tx| &mempool_tx.tx).collect()
+    }
+
+    fn contains_transaction(&self, tx_id: &Id<Transaction>) -> bool {
+        self.store.txs_by_id.contains_key(&tx_id.get())
+    }
+
+    // TODO Consider returning an error
+    fn drop_transaction(&mut self, tx_id: &Id<Transaction>) {
+        self.store.drop_tx(tx_id);
+        self.chain_state.drop_tx(tx_id)
+    }
+
+    fn get_chain_state(&self) -> &C {
+        &self.chain_state
+    }
+}
+
+fn has_duplicate_entry<T>(iter: T) -> bool
+where
+    T: IntoIterator,
+    T::Item: Eq + std::hash::Hash,
+{
+    let mut uniq = HashSet::new();
+    iter.into_iter().any(move |x| !uniq.insert(x))
```

Comment:
This is a relatively controversial thing to do. I noticed you added `Hash` derive for transaction, and I came here to see what's going on. Checking whether a transaction is in the mempool should be done through the Ideable trait, not just hashing. This topic is still not over and I discussed the necessity to have a bigger discussion about this with Lukas last week. Anyhow, we shouldn't be using `std::hash::Hash` to solve this problem. Please use the `Id` retrieved from the `Ideable` trait for this, at least until this problem is solved.

Reply:
Done.
I will say there was a bit of pain around the Idable business.
The reason for `derive(Hash)` was to be able to stick OutPoints in a set and as keys in HashMaps.
Since we decided we don't want to `derive(Hash)`, I decided to replaced `HashSet<OutPoint>` with `HashMap<Id<OutPoint>, OutPoint>`.

This of course requires that `Id<OutPoint>` implement `Hash`. I tried to `derive(Hash)` for `Id<OutPoint>` but for some reason deriving `Hash` for `Id<T>` requires `T` to implement (which brings us back where we started). So I manually implemented Hash for Id<OutPoint>. But clippy doesn't like when one of `PartialEq`, `Hash` is derived and the other isn't. So I had to disable a lint. Not the end of the world, but a bit ugly, and I don't see why the compiler wants `T` to be `Hash` to be able to `derive(Hash)` for `Id<T>`